pennylane/grouping/pauli.py

```diff
@@ -0,0 +1,246 @@
+
+    # Check if pauli_1 and pauli_2 are the same; if so, the result is the Identity
+    if are_identical_pauli_words(pauli_1, pauli_2):
+        first_wire = list(wire_map.keys())[0]
+        return Identity(first_wire)
+
+    # Compute binary symplectic representations
+    pauli_1_binary = pauli_to_binary(pauli_1, wire_map=wire_map)
+    pauli_2_binary = pauli_to_binary(pauli_2, wire_map=wire_map)
+
+    bin_symp_1 = np.array([int(x) for x in pauli_1_binary])
+    bin_symp_2 = np.array([int(x) for x in pauli_2_binary])
+
+    # Shorthand for bitwise XOR of numpy arrays
+    pauli_product = bin_symp_1 ^ bin_symp_2
+
+    return binary_to_pauli(pauli_product, wire_map=wire_map)
+
+
+def pauli_mult_with_phase(pauli_1, pauli_2, wire_map=None):
+    r"""Multiply two Pauli words together, and return both their product as a Pauli word
+    and the global phase.
+
+    Two Pauli operations can be multiplied together by taking the additive
+    OR of their binary symplectic representations. The phase is computed by
+    looking at the number of times we have the products  :math:`XY, YZ`, or :math:`ZX` (adds a
+    phase of :math:`i`), or :math:`YX, ZY, XZ` (adds a phase of :math:`-i`).
+
+    Args:
+        pauli_1 (.Operation): A Pauli word.
+        pauli_2 (.Operation): A Pauli word to multiply with the first one.
+        wire_map  (dict[Union[str, int], int]): dictionary containing all wire labels used in the Pauli
+            word as keys, and unique integer labels as their values. If no wire map is
+            provided, the map will be constructed from the set of wires acted on
+            by the input Pauli words.
+
+    Returns:
+        tuple[.Operation, complex]: The product of ``pauli_1`` and ``pauli_2``, and the
+        global phase.
+
+    **Example**
+
+    This function works the same as :func:`~.pauli_mult` but also returns the global
+    phase accumulated as a result of the ordering of Paulis in the product (e.g., :math:`XY = iZ`,
+    and :math:`YX = -iZ`).
+
+    >>> from pennylane.grouping import pauli_mult_with_phase
+    >>> pauli_1 = qml.PauliX(0) @ qml.PauliZ(1)
+    >>> pauli_2 = qml.PauliY(0) @ qml.PauliZ(1)
+    >>> product, phase = pauli_mult_with_phase(pauli_1, pauli_2)
+    >>> product
+    PauliZ(wires=[0])
+    >>> phase
+    1j
+    """
+
+    if wire_map is None:
+        wire_map = _wire_map_from_pauli_pair(pauli_1, pauli_2)
+
+    # Get the product; use our earlier function
+    pauli_product = pauli_mult(pauli_1, pauli_2, wire_map)
+
+    # Get the names of the operations; in cases where only one single-qubit Pauli
+    # is present, the operation name is stored as a string rather than a list, so convert it
+    pauli_1_string = pauli_word_to_string(pauli_1, wire_map=wire_map)
+    pauli_2_string = pauli_word_to_string(pauli_2, wire_map=wire_map)
+
+    pos_phases = (("X", "Y"), ("Y", "Z"), ("Z", "X"))
+
+    phase = 1
+
+    for qubit_1_char, qubit_2_char in zip(pauli_1_string, pauli_2_string):
```

Comment:
This function has been streamlined from the first iteration by converting the Pauli words into strings, and using the strings as a means to compare. This automatically handles any issues that might otherwise occur with qubits in the tensor product that are ordered differently than they are in the wire map; the `pauli_word_to_string` function takes care of that internally by virtue of ordering them and then inserting `"I"` on the unused wires.